src/bin/communication_thread.rs

```diff
@@ -0,0 +1,65 @@
+//! Example on how to use a communication thread alongside with the GUI thread.
+//! 
+//! Tricks used here:
+//! - Use a queue to show data on the GUI.
+//! - Run an async function on the GUI event loop.
+//! - Use a seperate thread to handle incoming data and put it into a queue.
+
+extern crate futures;
+extern crate gio;
+extern crate glib;
+extern crate gtk;
+use futures::StreamExt;
+use gio::prelude::*;
+use gtk::prelude::*;
+use gtk::{ApplicationWindow, Label};
+use std::env::args;
+use std::thread;
+
+fn main() {
+    let application = gtk::Application::new(
+        Some("com.github.gtk-rs.examples.communication_thread"),
+        Default::default(),
+    )
+    .expect("Initialization failed...");
+    application.connect_activate(build_ui);
+    application.run(&args().collect::<Vec<_>>());
+}
+
+fn build_ui(application: &gtk::Application) {
+    let window = ApplicationWindow::new(application);
+    let label = Label::new(None);
+    window.add(&label);
+
+    // Create a channel between communication thread and main event loop:
+    let (sender, mut receiver) = futures::channel::mpsc::channel::<String>(1000);
+
+    // Spawn queue receive task on the main event loop:
+    let main_context = glib::MainContext::default();
+    let future = async move {
+        while let Some(item) = receiver.next().await {
+            label.set_text(&item);
+        }
+    };
+    main_context.spawn_local(future);
+
+    start_communication_thread(sender);
+
+    window.show_all();
+}
+
+fn start_communication_thread(mut sender: futures::channel::mpsc::Sender<String>) {
+    // Spawn separate thread to handle communication:
+    thread::spawn(move || {
+        let mut counter = 0;
+        loop {
+            // Instead of a counter, your application code will
+            // block here on TCP or serial communications.
```

Comment:
Ideally you'd do such things asynchronously though, unless your APIs or use-case does not allow for that :) Maybe a comment about that could be useful here that threads are not always the best choice for such things.

Reply:
Agreed, I'll make a comment about it.